### PR TITLE
fix(core): add ariaLabelledby to the popover header for popover component and adding Tabindex to the list items

### PR DESCRIPTION
**Pull request by Lokanathan-k**
#### Please provide a link to the associated issue.
Closes: https://github.com/SAP/fundamental-ngx/issues/3143

#### Please provide a brief summary of this pull request.

add ariaLabelledby to the dialog header for popover component to handle accessibility and work along and introduced the initfocus to focus the click me button 

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [NA] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [NA] update `README.md`
- [NA] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Review by puru-hk — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <button fd-button label="Open Dialog" (click)="openDialog(template)"></button>
 
 <ng-template #template let-dialog let-dialogConfig="dialogConfig">
-    <fd-dialog [dialogConfig]="dialogConfig" [dialogRef]="dialog">
+    <fd-dialog [dialogConfig]="dialogConfig" [dialogRef]="dialog" role="dialog">
```
Comment (olshansky) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html:4]: Do you really need `role="dialog"` here? `dialog.component.html` has the same role as well, looks unnecessary

Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html:4]: yes, we need this without 'role="dialog"' the contents are read multiple times, with this header content is read only once

Comment (olshansky) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html:4]: @Lokanathan-k nice to know the reason, thank you!
So we need to raise a separate issue, which would describe a double role for dialog component _(for `<fd-dialog>` and inside fd-dialog template)_ and it needs investigating (and resolve if possible)

**Review by olshansky — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html (hunk 1 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-dialog/popover-dialog-example.component.html:4]: no need to raise issue, this issue resides in the Jaws Application for screen reading, on defining the role eliminates the duplicate reading.

**Review by Betrozov — APPROVED** [review]

**Review by sKudum — CHANGES_REQUESTED** [review]
Hi @Lokanathan-k , 

     Considering all the elements will give complete accessibility to this example or else it looks partially accessible. What do you say?

**Comment by puru-hk** [comment]
@Lokanathan-k There are other examples is not working for screen reader. Hence i have the change the label as required rework.

**Comment by Lokanathan-k** [comment]
> @Lokanathan-k There are other examples is not working for screen reader. Hence i have the change the label as required rework.

@puru-hk fixed please check

**Comment by sKudum** [comment]
Hi @Lokanathan-k ,

Using keyboard not able to access value help dialog button by which drop down is not showing up. Could you please have a look.

**Comment by Lokanathan-k** [comment]
> 
> 
> Hi @Lokanathan-k ,
> 
> Using keyboard not able to access value help dialog button by which drop down is not showing up. Could you please have a look.

@sKudum multi input does not allow the user to open the dropdown using the key press, once the drop down is opened we it allows the keyboard to access the items in the drop, Its the limitation of the multi input

**Review by sKudum — APPROVED** [review]

**Comment by puru-hk** [comment]
Looks fine

**Review by puru-hk — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 2]
```diff
@@ -15,10 +15,10 @@
     <fd-popover [focusTrapped]="true" [focusAutoCapture]="true">
         <fd-popover-control>
             <button fd-button label="With Header"></button>
-        </fd-popover-control>
+        </fd-popover-control>x
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:18]: accidental 'x' character added here

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 3]
```diff
@@ -15,10 +15,10 @@
...
         <fd-popover-body>
             <div fd-popover-body-header>
-                <div fd-bar [cozy]="true">
+                <div fd-bar [cozy]="true" tabindex="0">
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:21]: Why do we want the header to receive tab focus?

**Comment by InnaAtanasova** [comment]
Screen Shot 2021-06-25 at 4 46 06 PM
Screen Shot 2021-06-25 at 4 45 56 PM
Screen Shot 2021-06-25 at 4 45 52 PM

The following focus outlines should not have this styling.

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 3 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:21]: @mikerodonnell89 

the focus on key arrow down press we dont get the focus to the header part the focus is directed only to the list elements and not the header

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 4]
```diff
@@ -17,16 +17,12 @@
             <button fd-button label="With Header"></button>
         </fd-popover-control>
         <fd-popover-body>
-            <div fd-popover-body-header>
-                <div fd-bar [cozy]="true">
-                    <div fd-bar-left>
-                        <fd-bar-element>Cozy Header</fd-bar-element>
-                    </div>
-                </div>
-            </div>
```
Comment (InnaAtanasova) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:26]: Why do you remove this?

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 5]
```diff
@@ -17,16 +17,12 @@
...
             <ul fd-list>
+				<li fd-list-group-header>
+					<span fd-list-title>Cozy Header</span>
+				</li>
                 <li fd-list-item *ngFor="let option of list">
-                    <a [attr.href]="option.url" fd-list-title>{{ option.text }}</a>
+                    <span fd-list-title>{{ option.text }}</span>
```
Comment (InnaAtanasova) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:25]: I don't think this is solving the problem. You can't just change the link to a span.

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 4 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:26]: on using the fd-bar-element the focus in not directed to the header part so the change is required to migrate to list header.

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 5 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:25]: fixed

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 4 above)
Comment (InnaAtanasova) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:26]: I don't understand your answer. Please provide more info. Thank you

File: apps/docs/src/app/core/component-docs/popover/popover-docs.module.ts [hunk 6]
```diff
@@ -32,6 +32,7 @@ import { BarModule } from '@fundamental-ngx/core/bar';
 import { SegmentedButtonModule } from '@fundamental-ngx/core/segmented-button';
 import { FormModule } from '@fundamental-ngx/core/form';
 import { CheckboxModule } from '@fundamental-ngx/core/checkbox';
+import { LinkModule } from '@fundamental-ngx/core';
```
Comment (InnaAtanasova) [inline apps/docs/src/app/core/component-docs/popover/popover-docs.module.ts:35]: I don't think this is the correct way to import modules. Please use the new approach

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 4 above)
Comment (InnaAtanasova) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:26]: The header in this case is part of the Popover, it's not part of the List. You should be able to have a header without a List:
https://fundamental-styles.netlify.app/?path=/docs/components-popover--alignment

Screen Shot 2021-07-05 at 3 30 55 PM

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

**Comment by InnaAtanasova** [comment]
Also, the title of the PR mentions `dialog header` but it's not the case. Please correct accordingly. Thanks

**Comment by InnaAtanasova** [comment]
A few problems:
1. The subheader focus:
Screen Shot 2021-07-06 at 12 55 41 PM

2. List elements gets double focus:
Screen Shot 2021-07-06 at 12 55 55 PM
Screen Shot 2021-07-06 at 12 55 51 PM

3. Are you sure the Header and Subhader are read by the screen reader?

**Comment by Lokanathan-k** [comment]
> A few problems:
> 
> 1. The subheader focus: **yes, As per Accessibility team the  header item has to be read out.**
> 
> Screen Shot 2021-07-06 at 12 55 41 PM
> 
> 1. List elements gets double focus: **fixed**
> 
> Screen Shot 2021-07-06 at 12 55 55 PM
> 
> Screen Shot 2021-07-06 at 12 55 51 PM
> 
> 1. Are you sure the Header and Subhader are read by the screen reader?: **Yes**

Screenshot 2021-07-07 at 9 31 59 AM

**Comment by InnaAtanasova** [comment]
> > A few problems:
> > 
> > 1. The subheader focus: **yes, As per Accessibility team the  header item has to be read out.**
> > 
> > Screen Shot 2021-07-06 at 12 55 41 PM
> > 
> > 1. List elements gets double focus: **fixed**
> > 
> > Screen Shot 2021-07-06 at 12 55 55 PM
> > Screen Shot 2021-07-06 at 12 55 51 PM
> > 
> > 1. Are you sure the Header and Subhader are read by the screen reader?: **Yes**
> 
> Screenshot 2021-07-07 at 9 31 59 AM

:D yes, I know it has to be focussed but don't you find something wrong? The styling maybe?

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 3 above)
Comment (mikerodonnell89) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:21]: tabIndex="0" makes the element tabbable, there is nothing in the header we want tabbable, the user can do this: 

Screen Shot 2021-07-14 at 3 37 58 PM

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 7]
```diff
@@ -17,16 +17,16 @@
...
                     <div fd-bar-left>
                         <fd-bar-element>Cozy Header</fd-bar-element>
                     </div>
                 </div>
             </div>
-            <ul fd-list>
-                <li fd-list-item *ngFor="let option of list">
-                    <a [attr.href]="option.url" fd-list-title>{{ option.text }}</a>
+            <ul fd-list tabindex="-1">
+                <li fd-list-item *ngFor="let option of list" tabindex="0">
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:28]: Arrow keys are already working correctly here on main branch, no need for these changes

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
see comment

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 3 above)
Comment (mikerodonnell89) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:21]: tabindex="0" makes the element tabbable, there is nothing in the header we want tabbable, the user can do this:

Screen Shot 2021-07-14 at 4 26 43 PM

We should instead use aria roles to get the screenreader to read the content

**Comment by mikerodonnell89** [comment]
@Lokanathan-k wondering if there must be a better approach than to make elements that should not be tabbable, tabbable... I understand making an element tabbable will get it read, however there is nothing in the header we want tabbable, the user can do this:

Screen Shot 2021-07-14 at 4 26 43 PM

We should instead use aria roles to get the screenreader to read the content

**Comment by Lokanathan-k** [comment]
> @Lokanathan-k wondering if there must be a better approach than to use tabindex="0"... I understand making an element tabbable will get it read by the screenreader, however there is nothing in the header we want truly tabbable, because the user can do this:
> 
> Screen Shot 2021-07-14 at 4 26 43 PM
> 
> We should instead use aria roles to get the screenreader to read the content

>fixed the issue with aria-labelled by

**Review by InnaAtanasova — COMMENTED** [review]
The examples with List still get double focus:
Screen Shot 2021-07-21 at 2 22 36 PM
Screen Shot 2021-07-21 at 2 22 25 PM

**Comment by InnaAtanasova** [comment]
Screen Shot 2021-07-21 at 2 25 13 PM

Are you sure the Header and the subheader are read by the screen reader?

**Comment by InnaAtanasova** [comment]
Screen Shot 2021-07-21 at 2 28 01 PM

Why are both Header and Subheader with the same class (fd-popover__body-header)? Subheader should have different class. `.fd-popover__body-header` class adds rounded corners on the top and results in this:


Screen Shot 2021-07-21 at 2 28 07 PM

**Comment by Lokanathan-k** [comment]
> Screen Shot 2021-07-21 at 2 28 01 PM
> 
> Why are both Header and Subheader with the same class (fd-popover__body-header)? Subheader should have different class. `.fd-popover__body-header` class adds rounded corners on the top and results in this:
> 
> Screen Shot 2021-07-21 at 2 28 07 PM

fixed

**Comment by Lokanathan-k** [comment]
> The examples with List still get double focus:
> Screen Shot 2021-07-21 at 2 22 36 PM
> Screen Shot 2021-07-21 at 2 22 25 PM

fixed

**Comment by InnaAtanasova** [comment]
Screen Shot 2021-07-22 at 11 18 11 AM

I can still reproduce it in latest version of Chrome

**Comment by Lokanathan-k** [comment]
> Screen Shot 2021-07-22 at 11 18 11 AM
> 
> I can still reproduce it in latest version of Chrome

fixed

**Review by aberikashvili — APPROVED** [review]
LGTM

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]
Screen Shot 2021-07-29 at 2 24 30 PM

Focus problem

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 8]
```diff
@@ -6,7 +6,7 @@
         <fd-popover-body>
             <ul fd-list>
                 <li fd-list-item *ngFor="let option of list">
-                    <a [attr.href]="option.url" fd-list-title>{{ option.text }}</a>
+                    <a [attr.href]="option.url" fd-list-title tabindex="-1">{{ option.text }}</a>
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:9]: I think we need to find a way to do this automatically, so the dev doesn't have to enter this every time

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 8 above)
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:9]: Screenshot 2021-08-03 at 9 58 58 PM

by default <a> tag gets focus, to over ride the focus we can add tabindex in the fd-link at the host, but the font color changes as shown in the screen shot.

current implementation 
Screenshot 2021-08-03 at 10 02 37 PM

Comment (mikerodonnell89) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:9]: It could be done programmatically by drilling down through ContentChildren/ViewChildren and setting the tabindex manually.  A bit extreme but helps remove unnecessary code for the developer

**Comment by puru-hk** [comment]
Looks fine

**Review by puru-hk — APPROVED** [review]

**Review by sKudum — APPROVED** [review]

**Review by Betrozov — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html (hunk 8 above)
Comment (InnaAtanasova) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html:9]: In NGX the developer should not add `tabindex="-1"`, this should be handled by fund-ngx behind the scene.

File: apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html [hunk 9]
```diff
@@ -17,16 +17,16 @@
...
                     <div fd-bar-left>
                         <fd-bar-element>Cozy Header</fd-bar-element>
                     </div>
                 </div>
             </div>
-            <ul fd-list>
+            <ul fd-list aria-labelledby="popover-bar-header-1">
                 <li fd-list-item *ngFor="let option of list">
-                    <a [attr.href]="option.url" fd-list-title>{{ option.text }}</a>
+                    <a [attr.href]="option.url" fd-link fd-list-title tabindex="-1">{{ option.text }}</a>
```
[the remark this answers is not in the document]
Comment (Lokanathan-k) [reply apps/docs/src/app/core/component-docs/popover/examples/popover-complex-example/popover-complex-example.component.html]: @mikerodonnell89 @InnaAtanasova @aberikashvili fixed and screen reading is also fine
Screenshot 2021-08-06 at 6 48 22 PM

File: libs/core/src/lib/popover/popover.component.ts [hunk 10]
```diff
@@ -159,6 +166,16 @@ export class PopoverComponent extends BasePopoverClass implements AfterViewInit,
         if (this.popoverControl && this.triggers.includes('click')) {
             this.popoverControl.makeTabbable();
         }
+        this.listItems.forEach((li) => {
+            const listItem = this._getHtmlElement(li);
+            const listTitle = listItem.querySelector('.fd-list__title');
+            listTitle.setAttribute('tabindex', '-1');
+        });
+    }
```
Comment (InnaAtanasova) [inline libs/core/src/lib/popover/popover.component.ts:174]: does not work with ContentChildren/ViewChildren as Mike suggested?

Comment (Lokanathan-k) [reply libs/core/src/lib/popover/popover.component.ts:174]: It does uses the contentChildrens as the object to access the listItems, when adding the contentChildren to the directive(ListTitleDirective) the object is not created since there is not constructor defined in the directive, this throws a runtime exception. in order to avoid these i have used ListItemComponent and added the tab index.

Screenshot 2021-08-09 at 9 14 06 AM

**Comment by fkolar** [comment]
Let's keep it as it is now (tabindexes). Rather provide overall solution for both `core and platform`. When you look at the e.g. `core` in terms of `tabindex`, there is not much difference between solution from loki and what is already in master.

I would suggest, provide guideline for core  and for platform (here we have parent component for all of the components so it might be easier), when and how ot use tabindex and let's add it to the coding guideline, otherwise code in core and in platform is going to be inconsistent. 

- I would stick with binding ( in platform) as material has this
 - I would add it to the html referring this tabindex
 - For places where we know we need to turn it off the tabindex, we can write directly a value `tabindex=-1`
 
 I don't think we can find some solution that will do this magically behind the scenes. What do you think?

**Review by InnaAtanasova — APPROVED** [review]

**Comment by InnaAtanasova** [comment]
> Let's keep it as it is now (tabindexes). Rather provide overall solution for both `core and platform`. When you look at the e.g. `core` in terms of `tabindex`, there is not much difference between solution from loki and what is already in master.
> 
> I would suggest, provide guideline for core and for platform (here we have parent component for all of the components so it might be easier), when and how ot use tabindex and let's add it to the coding guideline, otherwise code in core and in platform is going to be inconsistent.
> 
> * I would stick with binding ( in platform) as material has this
> * I would add it to the html referring this tabindex
> * For places where we know we need to turn it off the tabindex, we can write directly a value `tabindex=-1`
> 
> I don't think we can find some solution that will do this magically behind the scenes. What do you think?

It is done with 5 lines of code. 6 days ago, just above my comments about handling it magically behind the scenes, @mikerodonnell89 explained what needs to be done:
Screen Shot 2021-08-10 at 12 42 56 PM

Today he added the code:
Screen Shot 2021-08-10 at 12 40 26 PM

**Comment by mikerodonnell89** [comment]
@Lokanathan-k @InnaAtanasova @fkolar  this is possible via ContentChild and no need to hardcode the list title classname, see this commit: https://github.com/SAP/fundamental-ngx/pull/5693/commits/684ac60bff4c9194278756fe118d2e30ddd391b6

However I think we should take it one step further and actually remove the tabIndex from the list title itself, if the list title is an anchor element, which I've done in this commit: https://github.com/SAP/fundamental-ngx/pull/5693/commits/bc1650885fcef546f54865a86791dd0a29e6c502

If anyone objects we can roll back the most recent commit I've linked

File: libs/core/src/lib/popover/popover.component.ts (hunk 10 above)
Comment (Lokanathan-k) [reply libs/core/src/lib/popover/popover.component.ts]: > @Lokanathan-k @InnaAtanasova @fkolar this is possible via ContentChild and no need to hardcode the list title classname, see this commit: [684ac60](https://github.com/SAP/fundamental-ngx/commit/684ac60bff4c9194278756fe118d2e30ddd391b6)
> 
> However I think we should take it one step further and actually remove the tabIndex from the list title itself, if the list title is an anchor element, which I've done in this commit: [bc16508](https://github.com/SAP/fundamental-ngx/commit/bc1650885fcef546f54865a86791dd0a29e6c502)
> 
> If anyone objects we can roll back the most recent commit I've linked

I had this solution before i had this introduced my changes as @droshev had shared comments in my Previous Pr's not to introduce changes in different component when the issue is pointed to the Specific Component. So i found the best alternative solution without introducing the change in the ListComponent.

Thank you, @mikerodonnell89 @InnaAtanasova Helping resolving the issue.

**Review by mikerodonnell89 — APPROVED** [review]

**Comment by droshev** [comment]
@mikerodonnell89 thanks
